pageserver/src/tenant.rs

```diff
@@ -1631,64 +1636,88 @@ impl Tenant {
                     info!("Tenant is already in Stopping state, skipping activation");
                 }
                 TenantState::Loading | TenantState::Attaching => {
-                    *current_state = TenantState::Active;
-
+                    *current_state = TenantState::Activating;
                     debug!(tenant_id = %self.tenant_id, "Activating tenant");
+                    activating = true;
+                    // Continue outside the closure. We need to grab timelines.lock()
+                    // and we plan to turn it into a tokio::sync::Mutex in a future patch.
+                }
+            }
+        });
+        if let Err(e) = result {
+            assert!(!activating, "transition into Activating is infallible");
+            return Err(e);
+        }
 
-                    let timelines_accessor = self.timelines.lock().unwrap();
-                    let not_broken_timelines = timelines_accessor
-                        .values()
-                        .filter(|timeline| timeline.current_state() != TimelineState::Broken);
+        if activating {
+            let timelines_accessor = self.timelines.lock().unwrap();
+            let not_broken_timelines = timelines_accessor
+                .values()
+                .filter(|timeline| timeline.current_state() != TimelineState::Broken);
 
-                    // Spawn gc and compaction loops. The loops will shut themselves
-                    // down when they notice that the tenant is inactive.
-                    tasks::start_background_loops(self.tenant_id);
+            // Spawn gc and compaction loops. The loops will shut themselves
+            // down when they notice that the tenant is inactive.
+            tasks::start_background_loops(self.tenant_id);
 
-                    let mut activated_timelines = 0;
+            let mut activated_timelines = 0;
```

Comment:
wouldn't this one will be always not_broken_timelines.size() ?

Reply:
In #4319 it just became infallible. Might be ok to such changes out of that PR series.

Reply:
Fixed in https://github.com/neondatabase/neon/pull/4319/commits/4001f441c06b9aee4c422133e08fbfa622449e51

Reply:
I spent some time trying to find this thread going back between first and fourth PRs.

Iterators are consumed in `for _ in iter` loops nor do they ever know the exact len, so 4001f44 had to go.